Introduce separate job status module with frontend to track job status

oliverguenther opened the pull request:
This PR introduces a new module, API and frontend that can track a job status by it's job UUID and act according to updates by the job itself, or by the ActiveSupport notifications on errors, cancellation, retrying, etc. of the job.

**TODOS**

- [x] Move `Delayed::Job::Status` introduced by table export options to its own module
- [x] Add `/api/v3/job_statuses/:uuid` endpoint
- [x] Make `CopyProjectJob` update the status with a redirection link to the copied target project
- [x] Generic status messages such as for error and success, when the job does not provide any special logic
- [x] Remove `max_attempts` from Delayed Job
  - [x] Add `retry_on` for the relevant jobs
- [x] Remove the export table specifics, and move them to the status modal instead
- [x] Fix remaining tests of  instantiate template functionality.
- [x] Extend project copying to redirect to job status

oliverguenther committed:
Add reference to user in status, extend enum, and add payload

- Adds the user reference directly to a job status, this avoids having a
separate object for many cases
- Adds a payload object to the status to transport information
- Extends the enum with a type when the job was cancelled

oliverguenther committed:
Allow jobs to define to store status even without a reference

oliverguenther committed:
Remove max_attempts from delayed job

From now on, jobs should use the AJ `retry_on` and `discard_on` methods
to decide on what exceptions it makes sense to retry.

oliverguenther committed:
Listen to the notifications in a separate listener class

oliverguenther committed:
Add some debugging information to the notifications

oliverguenther committed:
Remove job reference in status

oliverguenther committed:
Refactor statused jobs to test whether job is statused

oliverguenther committed:
Make copy project job store a status

oliverguenther committed:
Rewrite status as module

oliverguenther committed:
Add frontend for job status

oliverguenther committed:
Add API endpoint for job status

oliverguenther committed:
Lead user back when closing the page on job_statuses

oliverguenther committed:
Let CopyProjectJob redirect to the target project

oliverguenther committed:
Linting

oliverguenther committed:
Replace export modal with job status modal

openprojectci commented on frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts:
```diff
@@ -16,29 +16,21 @@ import {
- the actual download once the export is ready (delayed or not).
-
 The modal might also be used to only display the progress of an export. This will happen if a link for exporting is provided via the locals.
  */
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/373b673f1a4d1e1f6026ed300f93434bb30e713b/frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts#L28"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

oliverguenther committed:
Use onPush for table modal

oliverguenther committed:
Correctly upsert statuses in all cases

In case of tests, jobs never are running through the 'queued' step.
This causes jobs to be missing in later updates.

To avoid this, ensure that we just use upsert, which is faster anyway.

oliverguenther committed:
Set retry_on options for some jobs

oliverguenther committed:
Fix expected response in export

oliverguenther committed:
Fix export spec

oliverguenther committed:
Modify copy project to also redirect to job status

ulferts commented on app/controllers/work_packages_controller.rb:
```diff
@@ -83,12 +83,16 @@ def index
                      .result
 
-    redirect_to work_packages_export_path(export_storage.id)
+    if request.headers['Accept']&.include?('application/json')
```
Why not use a 

```
respond_to do |format|
  format.json { ...}
  format.html { ... }
```
?

oliverguenther replied:
That's because we override the format with the route constraint (e.g., `.xls` will result in excel mime type). I haven't found a way to differentiate between the request format and the accept header.

ulferts commented on config/initializers/delayed_job_config.rb:
```diff
@@ -37,5 +37,6 @@
-Delayed::Worker.max_attempts = 3
+# Do not retry jobs from delayed_job
+# instead use 'retry_on' activejob functionality
+Delayed::Worker.max_attempts = 1
```
Thank you :+1:

ulferts commented on frontend/src/app/modules/hal/services/hal-resource.service.ts:
```diff
@@ -307,7 +307,7 @@ export class HalResourceService {
 
   /**
-   * Get the hal type for an attribute.
+   * Get the hal type for an attributeex
```
Typo

ulferts commented on modules/job_status/app/controllers/job_status/job_statuses_controller.rb:
```diff
@@ -0,0 +1,52 @@
+
+      render_404
+    end
+  end
```
Just general thoughts:

The controller adds nothing specific to the application. We only need it for the route and to render the angular layout. The authorization is actually relevant for security, as the API will check the permission again.

Hopefully, we will have more and more need for controllers such as this as more and more of the application is moved to the front end.

Because of this, we should think about adding a generic `AngularController` that might be referenced from a variety of routes. There is already a controller such as this https://github.com/opf/openproject/blob/dev/modules/my_page/app/controllers/my_page/angular_controller.rb but it is buried in a domain specific module currently. The only actual differences we have to deal with is the menu we need to display (or not).

This is meant as a heads up to prepare us to make the change once the occasion arrises. No changes need now.

oliverguenther replied:
Good idea. I did not see why my page requires a 'no_page' layout since the controller does not set any project. I removed it in favor of the common angular service and it appears to work as expected

ulferts reviewed (changes requested):
Code looks mostly good. Only found some minor things. I like the layout of the solution.

When clicking through the code I was wondering about the general heading of the "export modal", which I know technically to be a general background job status modal now. But it would be nice to be able to tailor the modal to the context:
![image](https://user-images.githubusercontent.com/617519/86220021-4884e400-bb83-11ea-98fe-a676fe5d7e00.png)

When running the export and not having started the background jobs for some time. Once I started the background job, a 404 was returned to the user all of a sudden:
![image](https://user-images.githubusercontent.com/617519/86220440-e1b3fa80-bb83-11ea-9560-5ebdab58a274.png)

I haven't yet identified the source of this error precisely but it seems to occur when the job is deleted. Weirdly, the same request issued after a 404 returns a success afterwards.

The notification message on copying projects is hidden by the overlay:
![image](https://user-images.githubusercontent.com/617519/86223657-1d50c380-bb88-11ea-84f5-ec2cfd603884.png)
Copying projects runs into the same 404 as above on my machine.

The error also seems to occur if the request is issued when the job starts:

![image](https://user-images.githubusercontent.com/617519/86223935-7a4c7980-bb88-11ea-86dd-819529e22b46.png)

There might be a race condition there.

oliverguenther committed:
Add common angular controller for frontend rendering without anything

oliverguenther committed:
Add handling and spec in case of job status returning 404

oliverguenther committed:
Ensure we're not updating the user reference from delayed job

oliverguenther committed:
Fix not found text

ulferts reviewed (approved):
Looks really good now.

The minor issue of the notification being blurred by the overlay remains but that can be done separately. 

I believe that the change drastically improved the user flow when copying a project. This feels so much better.

I still feel that it would be desirable to have the modal be customized for the background job (heading and message) but we already agreed to do that in the next round.

Thanks for going the extra mile to reuse the `AngularController`. Hopefully, we will end up having only that one for the non API-part.
